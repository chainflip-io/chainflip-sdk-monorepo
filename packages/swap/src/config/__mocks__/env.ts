export default {
  RPC_NODE_HTTP_URL: 'http://rpc-node.test',
  START_HTTP_SERVICE: false,
  START_PROCESSOR: false,
  SWAPPING_APP_PORT: 8080,
  BITCOIN_DEPOSIT_TRACKER_URL: '',
  RPC_BROKER_HTTPS_URL: 'https://rpc-broker.test',
  RPC_NODE_WSS_URL: 'ws://rpc-node.test',
  CHAINFLIP_NETWORK: 'perseverance',
  QUOTE_TIMEOUT: 1000,
  NODE_ENV: 'test',
  CHAINALYSIS_API_KEY: undefined,
  INGEST_GATEWAY_URL: 'https://ingest-gateway.test',
  PROCESSOR_BATCH_SIZE: 50,
  PROCESSOR_TRANSACTION_TIMEOUT: 10_000,
  REDIS_URL: 'redis://localhost:6379',
  MAINTENANCE_MODE: false,
  FULLY_DISABLED_INTERNAL_ASSETS: new Set(),
  DISABLED_DEPOSIT_INTERNAL_ASSETS: new Set(),
  DISABLED_DESTINATION_INTERNAL_ASSETS: new Set(),
  MAX_CHANNELS_OPEN_PER_ADDRESS: 10,
  DISABLE_QUOTING: false,
  SOLANA_RPC_HTTP_URL: 'http://solana-rpc.test',
  BITCOIN_RPC_HTTP_URL: 'http://bitcoin-rpc.test',
  STABLE_COIN_SLIPPAGE_MIN_PRICE: 0.995,
  DCA_CHUNK_INTERVAL_BLOCKS: 2,
  QUOTER_BALANCE_TOLERANCE_PERCENT: 10,
  QUOTING_BASE_SLIPPAGE: {},
  QUOTER_USE_AUGMENT: true,
  BROKER_COMMISSION_BPS: 0,
  ON_CHAIN_DEFAULT_NETWORK_FEE_BPS: 5n,
  ON_CHAIN_STABLECOIN_NETWORK_FEE_BPS: 1n,
  MINIMUM_NETWORK_FEE_USDC: 500_000n,
};
