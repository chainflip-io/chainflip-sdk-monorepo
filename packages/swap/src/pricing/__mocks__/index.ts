import { ChainflipAsset, InternalAssetMap } from '@chainflip/utils/chainflip';

const prices: InternalAssetMap<number> & { Dot: number } = {
  Dot: 6.5,
  Usdt: 1,
  Usdc: 1,
  ArbUsdc: 1,
  SolUsdc: 1,
  Sol: 150,
  Btc: 65_000,
  Flip: 4,
  Eth: 2_000,
  ArbEth: 2_000,
  HubDot: 6.5,
  HubUsdc: 1,
  HubUsdt: 1,
};

export const getAssetPrice = async (asset: ChainflipAsset): Promise<number | undefined> =>
  prices[asset];
