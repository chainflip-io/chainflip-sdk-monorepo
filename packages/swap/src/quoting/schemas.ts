import { z } from 'zod';
import { numericString, unsignedInteger } from '@/shared/parsers';

export const marketMakerResponseSchema = z.union([
  z
    .object({
      id: z.string(),
      intermediate_amount: numericString,
      output_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      intermediateAmount: rest.intermediate_amount,
      outputAmount: rest.output_amount,
    })),
  z
    .object({
      id: z.string(),
      output_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      outputAmount: rest.output_amount,
    })),
]);
export type MarketMakerQuote = z.infer<typeof marketMakerResponseSchema>;

export const swapRateResponseSchema = z
  .object({
    // TODO: simplify when we know how Rust `Option` is encoded
    intermediary: unsignedInteger.optional().nullable(),
    output: unsignedInteger,
  })
  .transform((rate) => ({
    intermediateAmount: rate.intermediary?.toString(),
    outputAmount: rate.output.toString(),
  }));
export type BrokerQuote = z.infer<typeof swapRateResponseSchema>;
