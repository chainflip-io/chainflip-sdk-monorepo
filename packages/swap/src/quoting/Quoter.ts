import BigNumber from 'bignumber.js';
import { randomUUID } from 'crypto';
import { Subject, Subscription, filter } from 'rxjs';
import { Server } from 'socket.io';
import { swap, type SwapInput } from '@/amm-addon';
import { getPoolsNetworkFeeHundredthPips } from '@/shared/consts';
import { InternalAsset, assetConstants } from '@/shared/enums';
import { getHundredthPipAmountFromAmount } from '@/shared/functions';
import { SwapFee } from '@/shared/schemas';
import { QuotingSocket } from './authenticate';
import Leg from './Leg';
import {
  Leg as MarketMakerLeg,
  MarketMakerQuote,
  MarketMakerQuoteRequest,
  marketMakerResponseSchema,
} from './schemas';
import { Pool } from '../client';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { buildFee } from '../utils/fees';
import { handleExit } from '../utils/function';
import logger from '../utils/logger';
import { percentDifference } from '../utils/math';
import { getCachedPoolOrdersAndPrice } from '../utils/rpc';

export type QuoteType = 'pool' | 'market_maker';

type Quote = { marketMaker: string; quote: MarketMakerQuote };

const simulateSwap = async (input: SwapInput) => {
  const { swappedAmount } = await swap(input);

  return swappedAmount;
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(
    new BigNumber(amount)
      .times(price)
      .shiftedBy(assetConstants.Usdc.decimals - assetConstants[asset].decimals)
      .toFixed(0),
  );
};

export const differenceExceedsThreshold = (
  a: bigint,
  b: bigint,
  tolerance = env.QUOTE_APPROXIMATION_THRESHOLD,
) => percentDifference(a.toString(), b.toString()).gt(tolerance);

export default class Quoter {
  private readonly quotes$ = new Subject<Quote>();

  constructor(
    private readonly io: Server,
    private createId: () => string = randomUUID,
  ) {
    io.on('connection', (socket: QuotingSocket) => {
      logger.info(`market maker "${socket.data.marketMaker}" connected`);

      const cleanup = handleExit(() => {
        socket.disconnect();
      });

      socket.on('disconnect', () => {
        logger.info(`market maker "${socket.data.marketMaker}" disconnected`);
        cleanup();
      });

      socket.on('quote_response', (message) => {
        const result = marketMakerResponseSchema.safeParse(message);

        if (!result.success) {
          logger.warn(`received invalid quote response from "${socket.data.marketMaker}"`, {
            message,
            reason: result.error,
          });
          return;
        }

        this.quotes$.next({ marketMaker: socket.data.marketMaker, quote: result.data });
      });
    });
  }

  canQuote() {
    return env.USE_JIT_QUOTING && this.io.sockets.sockets.size > 0;
  }

  private async collectMakerQuotes(request: MarketMakerQuoteRequest): Promise<MarketMakerQuote[]> {
    const connectedClients = this.io.sockets.sockets.size;
    if (connectedClients === 0) return Promise.resolve([]);

    this.io.emit('quote_request', request);

    const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

    return new Promise((resolve) => {
      let sub: Subscription;

      let timer: ReturnType<typeof setTimeout>;

      const complete = () => {
        resolve([...clientsReceivedQuotes.values()]);
        sub.unsubscribe();
        clearTimeout(timer);
      };

      sub = this.quotes$
        .pipe(
          filter(
            ({ quote }) =>
              quote.request_id === request.request_id && quote.legs.length === request.legs.length,
          ),
        )
        .subscribe(({ marketMaker, quote }) => {
          clientsReceivedQuotes.set(marketMaker, quote);
          if (clientsReceivedQuotes.size === connectedClients) complete();
        });

      timer = setTimeout(complete, env.QUOTE_TIMEOUT);
    });
  }

  private async quoteLegs(legs: [Leg]): Promise<[bigint]>;
  private async quoteLegs(legs: [Leg, Leg]): Promise<[bigint, bigint]>;
  private async quoteLegs(legs: [Leg, Leg | null]): Promise<[bigint, bigint | null]>;
  private async quoteLegs([leg1, leg2]: [Leg] | [Leg, Leg | null]): Promise<
    [bigint] | [bigint, bigint | null]
  > {
    const requestId = this.createId();

    const requestLegs = [leg1.toMarketMakerJSON()] as
      | [MarketMakerLeg]
      | [MarketMakerLeg, MarketMakerLeg];

    if (leg2) requestLegs[1] = leg2.toMarketMakerJSON();

    const quoteRequest: MarketMakerQuoteRequest = { request_id: requestId, legs: requestLegs };

    const [quotes, firstPoolState, secondPoolState] = await Promise.all([
      this.collectMakerQuotes(quoteRequest),
      getCachedPoolOrdersAndPrice(leg1.getBaseAsset()),
      leg2 && getCachedPoolOrdersAndPrice(leg2.getBaseAsset()),
    ]);

    if (quotes.length === 0) throw new Error('no quotes received');

    const results = [
      simulateSwap(
        leg1.toSwapInput({
          limitOrders: quotes.flatMap(({ legs }) => legs[0]),
          ...firstPoolState,
        }),
      ),
    ] as [Promise<bigint>] | [Promise<bigint>, Promise<bigint> | null];

    results[1] = leg2
      ? simulateSwap(
          leg2.toSwapInput({
            limitOrders: quotes.flatMap(({ legs }) => legs[1]!),
            ...secondPoolState,
          }),
        )
      : null;

    return Promise.all(results);
  }

  async getQuote(
    srcAsset: InternalAsset,
    destAsset: InternalAsset,
    swapInputAmount: bigint,
    pools: Pool[],
  ) {
    const networkFee = getPoolsNetworkFeeHundredthPips(env.CHAINFLIP_NETWORK);

    let intermediateAmount: bigint | null = null;
    let outputAmount: bigint;
    let networkFeeUsdc: bigint | null = null;

    const fees: SwapFee[] = [];

    if (srcAsset === 'Usdc' || destAsset === 'Usdc') {
      if (srcAsset === 'Usdc') {
        networkFeeUsdc = getHundredthPipAmountFromAmount(swapInputAmount, networkFee);
      }

      const leg = Leg.of(srcAsset, destAsset, swapInputAmount - (networkFeeUsdc ?? 0n));
      [outputAmount] = await this.quoteLegs([leg]);

      if (destAsset === 'Usdc') {
        networkFeeUsdc = getHundredthPipAmountFromAmount(outputAmount, networkFee);
        outputAmount -= networkFeeUsdc!;
      }
    } else {
      let approximateUsdcAmount = await approximateIntermediateOutput(
        srcAsset,
        swapInputAmount.toString(),
      );

      networkFeeUsdc =
        approximateUsdcAmount && getHundredthPipAmountFromAmount(approximateUsdcAmount, networkFee);

      if (networkFeeUsdc && approximateUsdcAmount) {
        approximateUsdcAmount -= networkFeeUsdc;
      }

      const leg1 = Leg.of(srcAsset, 'Usdc', swapInputAmount);
      let leg2 = Leg.of('Usdc', destAsset, approximateUsdcAmount);

      const quotes = await this.quoteLegs([leg1, leg2]);

      if (
        quotes[1] === null ||
        differenceExceedsThreshold(approximateUsdcAmount! + networkFeeUsdc!, quotes[0])
      ) {
        networkFeeUsdc = getHundredthPipAmountFromAmount(quotes[0], networkFee);
        quotes[0] -= networkFeeUsdc;
        leg2 = Leg.of('Usdc', destAsset, quotes[0]);
        [quotes[1]] = await this.quoteLegs([leg2]);
      }

      const secondLegPoolFee = getHundredthPipAmountFromAmount(
        quotes[1],
        pools[1].liquidityFeeHundredthPips,
      );
      quotes[1] -= secondLegPoolFee;

      fees.push(buildFee(destAsset, 'LIQUIDITY', secondLegPoolFee));

      [intermediateAmount, outputAmount] = quotes;
    }

    fees.push(buildFee('Usdc', 'NETWORK', networkFeeUsdc!));

    return {
      intermediateAmount,
      outputAmount,
      includedFees: fees,
      quoteType: 'market_maker' as QuoteType,
    };
  }
}
