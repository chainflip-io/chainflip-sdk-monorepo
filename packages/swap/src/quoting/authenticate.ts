import * as crypto from 'crypto';
import type { Server } from 'socket.io';
import { promisify } from 'util';
import { z } from 'zod';
import { InternalAssetMap, getInternalAsset, InternalAsset, InternalAssets } from '@/shared/enums';
import { isNotNullish } from '@/shared/guards';
import { assetAndChain } from '@/shared/parsers';
import prisma from '../client';
import { type QuotingSocket } from './Quoter';
import baseLogger from '../utils/logger';

const logger = baseLogger.child({ module: 'quoter' });

const verifyAsync = promisify(crypto.verify);

type Middleware = Parameters<Server['use']>[0];
type Next = Parameters<Middleware>[1];

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

const mapAssets = (quotedAssets: InternalAsset[] | null): InternalAssetMap<boolean> => {
  assert(quotedAssets === null || quotedAssets.length !== 0, 'no assets quoted');

  return Object.fromEntries(
    Object.values(InternalAssets).map(
      (asset) => [asset, quotedAssets === null || quotedAssets.includes(asset)] as const,
    ),
  ) as InternalAssetMap<boolean>;
};

const authSchema = z.union([
  z
    .object({
      client_version: z.literal('1'),
      market_maker_id: z.string(),
      timestamp: z.number(),
      signature: z.string(),
    })
    .transform((data) => ({
      ...data,
      quoted_assets: mapAssets(null),
      account_id: data.market_maker_id,
    })),
  z.object({
    client_version: z.literal('2'),
    account_id: z.string(),
    timestamp: z.number(),
    signature: z.string(),
    quoted_assets: z
      .array(assetAndChain.transform(getInternalAsset))
      .transform((assets) => mapAssets(assets.filter(isNotNullish))),
  }),
]);

const parseKey = (key: string) => {
  try {
    return crypto.createPublicKey({ key: Buffer.from(key), format: 'pem', type: 'spki' });
  } catch {
    throw new Error('invalid public key');
  }
};

const authenticate = async (socket: QuotingSocket, next: Next) => {
  try {
    const result = authSchema.safeParse(socket.handshake.auth);

    logger.info('received auth', { auth: result.data, error: result.error?.message });

    assert(result.success, 'invalid auth');

    const auth = result.data;
    const timeElapsed = Date.now() - auth.timestamp;

    logger.info('time elapsed from handshake timestamp', { timeElapsed });

    assert(timeElapsed < 30_000 && timeElapsed >= -30_000, 'invalid timestamp');

    const marketMaker = await prisma.marketMaker.findUnique({
      where: { name: auth.account_id },
    });

    assert(marketMaker, 'market maker not found');

    const key = parseKey(marketMaker.publicKey);

    const signaturesMatch = await verifyAsync(
      null,
      Buffer.from(`${auth.account_id}${auth.timestamp}`, 'utf8'),
      key,
      Buffer.from(auth.signature, 'base64'),
    );

    assert(signaturesMatch, 'invalid signature');

    // https://socket.io/docs/v4/server-socket-instance/#socketdata
    // eslint-disable-next-line no-param-reassign
    socket.data = {
      marketMaker: marketMaker.name,
      quotedAssets: auth.quoted_assets,
      clientVersion: auth.client_version,
    };

    next();
  } catch (error) {
    logger.warn('authentication error', { error: (error as Error).message });
    next(error as Error);
  }
};

export default authenticate;
