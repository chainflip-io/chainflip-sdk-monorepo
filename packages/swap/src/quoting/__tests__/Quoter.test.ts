/* eslint-disable dot-notation */
import BigNumber from 'bignumber.js';
import * as crypto from 'crypto';
import { Server } from 'socket.io';
import { io, Socket } from 'socket.io-client';
import { setTimeout as sleep } from 'timers/promises';
import { promisify } from 'util';
import { assetConstants, getAssetAndChain } from '@/shared/enums';
import env from '@/swap/config/env';
import prisma, { InternalAsset, Pool } from '../../client';
import { getAssetPrice } from '../../pricing';
import authenticate from '../authenticate';
import Quoter, { approximateIntermediateOutput, differenceExceedsThreshold } from '../Quoter';
import { MarketMakerQuote, MarketMakerRawQuote } from '../schemas';

const generateKeyPairAsync = promisify(crypto.generateKeyPair);

const serializeBigInt = (k: string, v: unknown) => (typeof v === 'bigint' ? v.toString() : v);

jest.mock('../../utils/statechain', () => ({
  getSwapRate: jest.fn((args) =>
    Promise.reject(new Error(`unhandled getSwapRate(${JSON.stringify(args, serializeBigInt)})`)),
  ),
}));

jest.mock('../../pricing');

jest.mock('../PoolStateCache');

function toAtomicUnits(amount: number, asset: InternalAsset, output?: 'string'): string;
function toAtomicUnits(amount: number, asset: InternalAsset, output: 'bigint'): bigint;
function toAtomicUnits(amount: number, asset: InternalAsset, output: string | bigint = 'string') {
  const amt = new BigNumber(amount).shiftedBy(assetConstants[asset].decimals).toFixed(0);

  return output === 'string' ? amt : BigInt(amt);
}

describe(Quoter, () => {
  let oldEnv: typeof env;
  let quoter: Quoter;
  let connectClient: (name: string) => Promise<{
    sendQuote: (quote: MarketMakerRawQuote) => MarketMakerQuote;
    socket: Socket;
  }>;
  let server: Server;
  const sockets: Socket[] = [];

  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE private."MarketMaker" CASCADE`;
    oldEnv = { ...env };
    server = new Server().use(authenticate).listen(0);

    quoter = new Quoter(server);

    connectClient = async (name: string) => {
      const { publicKey, privateKey } = await generateKeyPairAsync('ed25519');
      await prisma.marketMaker.create({
        data: {
          name,
          publicKey: publicKey.export({ format: 'pem', type: 'spki' }).toString(),
        },
      });

      const { port } = server['httpServer'].address();
      const timestamp = Date.now();

      const socket = io(`http://localhost:${port}`, {
        auth: {
          client_version: '1',
          market_maker_id: name,
          timestamp,
          signature: crypto
            .sign(null, Buffer.from(`${name}${timestamp}`, 'utf8'), privateKey)
            .toString('base64'),
        },
      });

      await new Promise<void>((resolve, reject) => {
        sleep(500).then(() => reject(new Error('socket did not connect')));
        socket.on('connect', resolve);
      });

      sockets.push(socket);

      return {
        socket,
        sendQuote(quote: MarketMakerRawQuote) {
          socket.emit('quote_response', quote);

          expect(socket.connected).toBe(true);

          return {
            request_id: quote.request_id,
            legs: quote.legs.map((leg) => leg.map(([tick, amount]) => [tick, BigInt(amount)])),
          } as MarketMakerQuote;
        },
      };
    };
  });

  afterEach(async () => {
    Object.assign(env, oldEnv);
    await promisify(server.close.bind(server))();
    await Promise.all(
      sockets.splice(0).map(
        (socket) =>
          new Promise<unknown>((resolve) => {
            socket.on('disconnect', resolve);
            socket.disconnect();
          }),
      ),
    );
  });

  describe('constructor', () => {
    it('ignores malformed quote responses', () => {
      const fakeServer = { on: jest.fn() };

      const mockQuoter = new Quoter(fakeServer as unknown as Server);

      expect(fakeServer.on).toHaveBeenCalledWith('connection', expect.any(Function));

      const handler = fakeServer.on.mock.calls[0][1];

      const socket = { on: jest.fn(), data: { marketMaker: 'MM' } };
      const next = jest.fn();
      mockQuoter['quotes$'].subscribe(next);

      handler(socket as any);

      const callback = socket.on.mock.calls[1][1];

      callback({ request_id: 'string', legs: [[[-1, '123456']]] });
      callback({ request_id: 'string', range_orders: [] });

      expect(next).toHaveBeenCalledTimes(1);
    });
  });

  describe(Quoter.prototype['collectMakerQuotes'], () => {
    const collectQuotes = () => {
      const id = crypto.randomUUID();

      return {
        promise: quoter['collectMakerQuotes']({
          request_id: id,
          legs: [
            {
              base_asset: getAssetAndChain('Flip'),
              quote_asset: getAssetAndChain('Usdc'),
              amount: '1000',
              side: 'SELL',
            },
          ],
        }),
        id,
      };
    };

    it('returns an empty array if expectedQuotes is 0', async () => {
      expect(await quoter['collectMakerQuotes']({ request_id: 'id', legs: [] as any })).toEqual([]);
    });

    it('returns an empty array if no quotes are received', async () => {
      env.QUOTE_TIMEOUT = 10;
      await connectClient('marketMaker');
      const { promise } = collectQuotes();
      expect(await promise).toEqual([]);
    });

    it('returns an array of quotes if expectedQuotes is received', async () => {
      env.QUOTE_TIMEOUT = 10_000;
      const { sendQuote } = await connectClient('marketMaker');
      const { id, promise } = collectQuotes();
      const quote = sendQuote({ request_id: id, legs: [[[0, '100']]] });
      expect(await promise).toEqual([quote]);
    });

    it('accepts the most recent quote from each market maker', async () => {
      env.QUOTE_TIMEOUT = 10;
      const { sendQuote } = await connectClient('marketMaker');
      await connectClient('marketMaker2');
      const { id, promise } = collectQuotes();
      sendQuote({ request_id: id, legs: [[[0, '100']]] });
      const quote = sendQuote({ request_id: id, legs: [[[0, '200']]] });
      expect(await promise).toEqual([quote]);
    });

    it.each([10, 50, 100])('can be configured with QUOTE_TIMEOUT', async (timeout) => {
      env.QUOTE_TIMEOUT = timeout;
      const { sendQuote } = await connectClient('marketMaker');
      const { id, promise } = collectQuotes();
      await sleep(timeout + 1);
      sendQuote({ request_id: id, legs: [[[0, '100']]] });
      expect(await promise).toEqual([]);
    });

    it('eagerly returns after all expected quotes are received', async () => {
      env.QUOTE_TIMEOUT = 10_000;
      const mm1 = await connectClient('marketMaker');
      const mm2 = await connectClient('marketMaker2');
      const { id, promise } = collectQuotes();
      const quote1 = mm1.sendQuote({ request_id: id, legs: [[[0, '100']]] });
      const quote2 = mm2.sendQuote({ request_id: id, legs: [[[0, '200']]] });
      // no need to advance timers because setTimeout is never called
      expect(await promise).toEqual([quote2, quote1]);
    });
  });

  describe(Quoter.prototype.getQuote, () => {
    const ONE_USDC = toAtomicUnits(1, 'Usdc', 'bigint');
    const ONE_FLIP = toAtomicUnits(1, 'Flip', 'bigint');
    const ONE_BTC = toAtomicUnits(1, 'Btc', 'bigint');
    const ONE_DOT = toAtomicUnits(1, 'Dot', 'bigint');
    const ONE_USDT = toAtomicUnits(1, 'Usdt', 'bigint');

    it('gets the quote for a single leg (USDC => FLIP)', async () => {
      quoter['createId'] = () => 'id';
      const { sendQuote } = await connectClient('marketMaker');
      const quote = quoter.getQuote('Usdc', 'Flip', ONE_USDC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
      ]);
      sendQuote({ request_id: 'id', legs: [[[-260483, ONE_FLIP.toString()]]] });

      expect(await quote).toEqual({
        intermediateAmount: null,
        outputAmount: 204942885815213259n,
        includedFees: [
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '1000',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for a single leg (FLIP => USDC)', async () => {
      quoter['createId'] = () => 'id';
      const { sendQuote } = await connectClient('marketMaker');
      const quote = quoter.getQuote('Flip', 'Usdc', ONE_FLIP, [
        { liquidityFeeHundredthPips: 0 } as Pool,
      ]);
      sendQuote({ request_id: 'id', legs: [[[-260483, (ONE_USDC * 5n).toString()]]] });

      expect(await quote).toEqual({
        intermediateAmount: null,
        outputAmount: 4869653n,
        includedFees: [
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '4874',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for a single leg (BTC => USDC)', async () => {
      quoter['createId'] = () => 'id';
      const { sendQuote } = await connectClient('marketMaker');
      const quote = quoter.getQuote('Btc', 'Usdc', ONE_BTC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
      ]);
      sendQuote({ request_id: 'id', legs: [[[65636, (ONE_USDC * 75_000n).toString()]]] });

      expect(await quote).toEqual({
        intermediateAmount: null,
        outputAmount: 70787770203n,
        includedFees: [
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '70858628',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for two legs (BTC => USDC => FLIP)', async () => {
      jest.mocked(getAssetPrice).mockResolvedValueOnce(70_000.0);
      const ids = ['id2', 'id1'];
      quoter['createId'] = () => ids.pop() as string;
      const { sendQuote, socket } = await connectClient('marketMaker');

      const quotes: Record<string, MarketMakerRawQuote> = {
        id1: {
          request_id: 'id1',
          legs: [
            [[65636, (ONE_USDC * 75_000n).toString()]], // BTC => USDC
            [[-260483, (ONE_FLIP * 15_000n).toString()]], // USDC => FLIP
          ],
        },
        id2: {
          request_id: 'id2',
          legs: [
            [[-260483, (ONE_FLIP * 15_000n).toString()]], // USDC => FLIP
          ],
        },
      };

      socket.on('quote_request', (req) => {
        sendQuote(quotes[req.request_id]);
      });

      const quote = quoter.getQuote('Btc', 'Flip', ONE_BTC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
        { liquidityFeeHundredthPips: 1000 } as Pool,
      ]);

      expect(await quote).toEqual({
        intermediateAmount: 70787770203n,
        outputAmount: 14507449905826984626132n,
        includedFees: [
          {
            amount: '14521971877704689315',
            asset: 'FLIP',
            chain: 'Ethereum',
            type: 'LIQUIDITY',
          },
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '70858628',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for two legs (DOT => USDC => USDT)', async () => {
      jest.mocked(getAssetPrice).mockResolvedValueOnce(6.5);
      const ids = ['id2', 'id1'];
      quoter['createId'] = () => ids.pop() as string;
      const { sendQuote, socket } = await connectClient('marketMaker');

      const quotes: Record<string, MarketMakerRawQuote> = {
        id1: {
          request_id: 'id1',
          legs: [
            [[-73540, (ONE_USDC * 65_000n).toString()]], // DOT => USDC
            [[0, (ONE_USDT * 65_000n).toString()]], // USDC => USDT
          ],
        },
        id2: {
          request_id: 'id2',
          legs: [
            [[0, (ONE_USDT * 65_000n).toString()]], // USDC => USDT
          ],
        },
      };

      socket.on('quote_request', (req) => {
        sendQuote(quotes[req.request_id]);
      });

      const quote = quoter.getQuote('Dot', 'Usdt', ONE_DOT * 10_000n, [
        { liquidityFeeHundredthPips: 0 } as Pool,
        { liquidityFeeHundredthPips: 0 } as Pool,
      ]);

      expect(await quote).toEqual({
        includedFees: [
          {
            amount: '0',
            asset: 'USDT',
            chain: 'Ethereum',
            type: 'LIQUIDITY',
          },
          {
            amount: '64026249',
            asset: 'USDC',
            chain: 'Ethereum',
            type: 'NETWORK',
          },
        ],
        intermediateAmount: 63962222785n,
        outputAmount: 63962222784n,
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for two legs in one round trip (BTC => USDC => FLIP)', async () => {
      jest.mocked(getAssetPrice).mockResolvedValueOnce(70_787.0);
      quoter['createId'] = () => 'id';
      const { sendQuote } = await connectClient('marketMaker');
      const quote = quoter.getQuote('Btc', 'Flip', ONE_BTC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
        { liquidityFeeHundredthPips: 1000 } as Pool,
      ]);
      sendQuote({
        request_id: 'id',
        legs: [
          [[65636, (ONE_USDC * 75_000n).toString()]], // BTC => USDC
          [[-260483, (ONE_FLIP * 15_000n).toString()]], // USDC => FLIP
        ],
      });

      expect(await quote).toEqual({
        intermediateAmount: 70858628831n,
        outputAmount: 14492784766143299590341n,
        includedFees: [
          {
            amount: '14507292058201501091',
            asset: 'FLIP',
            chain: 'Ethereum',
            type: 'LIQUIDITY',
          },
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '70787000',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('gets the quote for two legs without a USDC estimate (BTC => USDC => FLIP)', async () => {
      jest.mocked(getAssetPrice).mockResolvedValueOnce(undefined);
      const ids = ['id2', 'id1'];
      quoter['createId'] = () => ids.pop() as string;
      const { sendQuote, socket } = await connectClient('marketMaker');

      const quotes: Record<string, MarketMakerRawQuote> = {
        id1: {
          request_id: 'id1',
          legs: [
            [[65636, (ONE_USDC * 75_000n).toString()]], // BTC => USDC
          ],
        },
        id2: {
          request_id: 'id2',
          legs: [
            [[-260483, (ONE_FLIP * 15_000n).toString()]], // USDC => FLIP
          ],
        },
      };

      socket.on('quote_request', (req) => {
        sendQuote(quotes[req.request_id]);
      });

      const quote = quoter.getQuote('Btc', 'Flip', ONE_BTC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
        { liquidityFeeHundredthPips: 1000 } as Pool,
      ]);

      expect(await quote).toEqual({
        intermediateAmount: 70787770203n,
        outputAmount: 14507449905826984626132n,
        includedFees: [
          {
            amount: '14521971877704689315',
            asset: 'FLIP',
            chain: 'Ethereum',
            type: 'LIQUIDITY',
          },
          {
            type: 'NETWORK',
            chain: 'Ethereum',
            asset: 'USDC',
            amount: '70858628',
          },
        ],
        quoteType: 'market_maker',
      });
    });

    it('throws if no quotes are received', async () => {
      await connectClient('marketMaker');

      const quote = quoter.getQuote('Usdc', 'Flip', ONE_USDC, [
        { liquidityFeeHundredthPips: 0 } as Pool,
      ]);

      await expect(quote).rejects.toThrow('no quotes received');
    });
  });
});

describe(approximateIntermediateOutput, () => {
  it.each([
    ['Dot', 6.5],
    ['Usdt', 1],
    ['Btc', 65_000],
    ['Flip', 4],
    ['Eth', 2_000],
  ] as [InternalAsset, number][])(
    'correctly approximates the USD value of %s',
    async (asset, price) => {
      jest.mocked(getAssetPrice).mockResolvedValueOnce(price);

      const oneInAtomicUnits = toAtomicUnits(1, asset);

      const actual = await approximateIntermediateOutput(asset, oneInAtomicUnits);

      const expected = toAtomicUnits(price, 'Usdc', 'bigint');

      expect(actual).toEqual(expected);
    },
  );
});

describe(differenceExceedsThreshold, () => {
  it('returns false if the difference does not exceed the threshold', () => {
    expect(differenceExceedsThreshold(100n, 101n, 1)).toBe(false);
  });

  it('returns false if the difference exceeds the threshold', () => {
    expect(differenceExceedsThreshold(100n, 102n, 1)).toBe(true);
  });
});
