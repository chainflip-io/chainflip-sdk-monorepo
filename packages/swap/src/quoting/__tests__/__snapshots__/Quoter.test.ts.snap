// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Quoter > getLimitOrders > can mix and match > flip 1`] = `
{
  "legs": [
    {
      "amount": "15000000",
      "base_asset": {
        "asset": "FLIP",
        "chain": "Ethereum",
      },
      "quote_asset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
      "side": "BUY",
    },
  ],
  "request_id": Any<String>,
}
`;

exports[`Quoter > getLimitOrders > can mix and match > solana 1`] = `
{
  "legs": [
    {
      "amount": "100000000",
      "base_asset": {
        "asset": "SOL",
        "chain": "Solana",
      },
      "quote_asset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
      "side": "SELL",
    },
  ],
  "request_id": Any<String>,
}
`;

exports[`Quoter > getLimitOrders > can quote both legs 1`] = `
{
  "legs": [
    {
      "amount": "100000000",
      "base_asset": {
        "asset": "SOL",
        "chain": "Solana",
      },
      "quote_asset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
      "side": "SELL",
    },
    {
      "amount": "15000000",
      "base_asset": {
        "asset": "FLIP",
        "chain": "Ethereum",
      },
      "quote_asset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
      "side": "BUY",
    },
  ],
  "request_id": Any<String>,
}
`;

exports[`Quoter > getLimitOrders > can quote the first leg only 1`] = `
{
  "legs": [
    {
      "amount": "100000000",
      "base_asset": {
        "asset": "SOL",
        "chain": "Solana",
      },
      "quote_asset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
      "side": "SELL",
    },
  ],
  "request_id": Any<String>,
}
`;
